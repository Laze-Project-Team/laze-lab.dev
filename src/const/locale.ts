import type { localeId } from '@/lib/utils/isLocale';

export const localeList: Record<localeId, string> = {
  ja: '日本語',
  en: 'English',
};
