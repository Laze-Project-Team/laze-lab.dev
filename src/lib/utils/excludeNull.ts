type excludeNullType<T> = Exclude<T, null | undefined>;

export const excludeNull = <T>(arr: T[]): excludeNullType<T>[] => {
  return arr.filter(
    (val) => val !== null && val !== undefined,
  ) as excludeNullType<T>[];
};
