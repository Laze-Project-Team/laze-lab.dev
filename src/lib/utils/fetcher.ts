export const fetcher = (
  input: RequestInfo | URL,
  init?: RequestInit,
): unknown => fetch(input, init).then((res) => res.json());
