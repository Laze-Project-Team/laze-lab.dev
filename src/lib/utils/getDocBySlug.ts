import fs from 'fs';
import matter from 'gray-matter';
import path from 'path';

import { contentsDir } from '@/lib/path';

type docMeta = {
  slug: string;
  meta: {
    title: string;
    description: string;
  };
  content: string;
};

const isMeta = (data: Record<string, unknown>): data is docMeta['meta'] =>
  typeof data.title === 'string' && typeof data.description === 'string';

export const getDocBySlug = (slug: string): docMeta => {
  const fileContents = fs.readFileSync(path.join(contentsDir, slug), 'utf-8');
  const { content, data } = matter(fileContents);

  if (!isMeta(data)) {
    throw new Error(`missing title or description in "${slug}"`);
  }

  return { slug, meta: data, content };
};
