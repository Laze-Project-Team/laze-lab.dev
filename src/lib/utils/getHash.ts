export const getHash = (): string =>
  Date.now().toString(36) + Math.random().toString(36);
