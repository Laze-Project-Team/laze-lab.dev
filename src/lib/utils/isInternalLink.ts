import type { UrlObject } from 'url';

const isInternalHrefString = (href: string): boolean =>
  href.startsWith('/') || href.startsWith('#');

export const isInternalLink = (href: UrlObject | string): boolean => {
  if (typeof href === 'string') {
    return isInternalHrefString(href);
  } else {
    const pathname = href.pathname;

    // if null/undefined
    if (pathname == undefined) return false;

    return isInternalHrefString(pathname);
  }
};
