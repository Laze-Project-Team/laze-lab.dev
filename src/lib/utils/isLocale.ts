export const locales = ['ja', 'en'] as const;
export type localeId = (typeof locales)[number];

export const isLocale = (locale: unknown): locale is localeId => {
  if (typeof locale !== 'string') return false;

  return locales.includes(locale as localeId);
};
