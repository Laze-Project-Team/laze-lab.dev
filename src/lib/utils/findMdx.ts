import fs from 'fs';
import path from 'path';

import { contentsDir } from '@/lib/path';
import { excludeNull } from '@/lib/utils/excludeNull';
import type { localeId } from '@/lib/utils/isLocale';
import { isLocale } from '@/lib/utils/isLocale';

const getFullPath = (pathname: string) => path.join(contentsDir, pathname);

type mdxType = {
  params: {
    path: string;
  };
  locale: localeId;
};

export const findMdx = (pathname: string): mdxType[] => {
  const dirents = fs.readdirSync(getFullPath(pathname), {
    withFileTypes: true,
  });
  const paths = excludeNull(
    dirents.map((dirent) => {
      if (!dirent.isFile()) return;
      if (!dirent.name.endsWith('.mdx')) return;

      const match = dirent.name.match(/([^.]*).([^.]*).mdx/);
      if (match === null) {
        throw new Error(
          `locale is not specified at "${path.join(pathname, dirent.name)}"`,
        );
      }

      const name = match[1];
      const locale = match[2];
      if (!isLocale(locale)) {
        throw new Error(
          `locale at "${path.join(pathname, dirent.name)}" is not valid`,
        );
      }

      const mdx: mdxType = {
        params: { path: name },
        locale,
      };

      return mdx;
    }),
  );

  return paths;
};

export const findMdxRecursively = (pathname: string): mdxType[] => {
  const dirents = fs.readdirSync(getFullPath(pathname), {
    withFileTypes: true,
  });

  const paths = excludeNull(
    dirents.flatMap((dirent) => {
      if (!dirent.isDirectory()) return;

      return findMdx(path.join(pathname, dirent.name));
    }),
  );

  return findMdx(pathname).concat(paths);
};
