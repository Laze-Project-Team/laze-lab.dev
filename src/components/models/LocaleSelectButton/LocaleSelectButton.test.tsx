import '@testing-library/jest-dom';

import { fireEvent, screen } from '@testing-library/react';
import { expect, it } from 'vitest';

import { mockEachRouter } from '/__mocks__/router';
import { localeList } from '@/const/locale';
import { render } from '@/lib/test/render';
import { locales } from '@/lib/utils/isLocale';

import { LocaleSelectButton } from '.';

const router = mockEachRouter();

it('should be rendered correctly', () => {
  const { container } = render(<LocaleSelectButton />);
  expect(container.firstChild).toMatchSnapshot();
});

it('should have "button" role', () => {
  render(<LocaleSelectButton />);
  expect(screen.getByRole('button')).toBeInTheDocument();
});

it('can change locale', async () => {
  render(<LocaleSelectButton />);
  const button = screen.getByRole('button');

  await Promise.all(
    locales.map(async (locale) => {
      fireEvent.mouseDown(button);
      const menuItem = await screen.findByRole('option', {
        name: localeList[locale],
      });

      fireEvent.click(menuItem);
      expect(button).toContainHTML(localeList[locale]);

      expect(router.push).toHaveBeenCalledWith(router.pathname, router.asPath, {
        locale,
      });
    }),
  );
});
