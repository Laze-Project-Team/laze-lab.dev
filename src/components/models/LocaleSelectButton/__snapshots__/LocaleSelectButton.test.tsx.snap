// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should be rendered correctly 1`] = `
<div
  class="MuiFormControl-root css-8zhps8-MuiFormControl-root-PresentialLocaleSelectButton"
>
  <label
    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
    data-shrink="true"
    id="locale-select-label"
  >
    changeLocale
  </label>
  <div
    class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
  >
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-labelledby="locale-select-label locale-select"
      class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-jedpe8-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
      id="locale-select"
      role="button"
      tabindex="0"
    >
      English
    </div>
    <input
      aria-hidden="true"
      class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
      defaultvalue="en"
      tabindex="-1"
    />
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
      data-testid="ArrowDropDownIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M7 10l5 5 5-5z"
      />
    </svg>
    <fieldset
      aria-hidden="true"
      class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        class="css-14lo706"
      >
        <span>
          changeLocale
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;
